aim/db/hashtree_db_listener.py

```diff
@@ -179,23 +179,22 @@ def _retrieve_class_root_type(self, klass, cache=None):
 
     def catch_up_with_action_log(self, store, served_tenants=None):
         ctx = utils.FakeContext(store=store)
-        with ctx.store.begin(subtransactions=True):
-            served_tenants = served_tenants or set()
-            to_init = set(self.tt_mgr.retrieve_uninitialized_roots(ctx))
-            served_tenants |= to_init
-            # Nothing will happen if there's no action log
-            kwargs = {'order_by': ['root_rn', 'id']}
-            if served_tenants:
-                kwargs['in_'] = {'root_rn': served_tenants}
-            logs = self.aim_manager.find(ctx, aim_tree.ActionLog, **kwargs)
-            LOG.debug('Processing action logs: %s' % logs)
-            log_by_root, resetting_roots = self._preprocess_logs(ctx, logs)
-            self._cleanup_resetting_roots(ctx, log_by_root, resetting_roots)
-            self._push_changes_to_trees(ctx, log_by_root)
-            # REVISIT: This is temporary code for verifying solutions
-            # to concurrency issues. Remove when no longer needed.
-            if aim_cfg.CONF.aim.validate_config_trees:
-                self._validate_config_trees(ctx, log_by_root.keys())
+        served_tenants = served_tenants or set()
```

Comment:
I think we want to keep an outer transaction, but we should instead scope it by each tenant. This has the downside of more DB round trips (one per-tenant, instead of one for all tenants), but ensures that any operations for a tenant related to action logs can still be rolled-back (and therefore retried) if there's an exception.